doc/extdev/event_callbacks.rst

```diff
@@ -0,0 +1,406 @@
+   # or the doctree-resolved event is emitted
+   for docname in updated-docs:
+      13. apply post-transforms (by priority): docutils.document -> docutils.document
+      14. event.doctree-resolved(app, doctree, docname)
+          - In the event that any reference nodes fail to resolve, the following may emit:
+          - event.missing-reference(env, node, contnode)
+          - event.warn-missing-reference(domain, node)
+
+   15. Generate output files
+   16. event.build-finished(app, exception)
+
+Here is also a flow diagram of the events,
+within the context of the Sphinx build process:
+
+.. graphviz:: /_static/diagrams/sphinx_core_events_flow.dot
+   :caption: Sphinx core events flow
+
+Core event details
+------------------
+
+Here is a more detailed list of these events.
+
+.. event:: config-inited (app, config)
+
+   :param app: :class:`.Sphinx`
+   :param config: :class:`.Config`
+
+   Emitted when the config object has been initialized.
+
+   .. versionadded:: 1.8
+
+.. event:: builder-inited (app)
+
+   :param app: :class:`.Sphinx`
+
+   Emitted when the builder object has been created
+   (available as ``app.builder``).
+
+.. event:: env-get-outdated (app, env, added, changed, removed)
+
+   :param app: :class:`.Sphinx`
+   :param env: :class:`.BuildEnvironment`
+   :param added: ``set[str]``
+   :param changed: ``set[str]``
+   :param removed: ``set[str]``
+   :returns: ``list[str]`` of additional docnames to re-read
+
+   Emitted when the environment determines which source files have changed and
+   should be re-read.
+   *added*, *changed* and *removed* are sets of docnames
+   that the environment has determined.
+   You can return a list of docnames to re-read in addition to these.
+
+   .. versionadded:: 1.1
+
+.. event:: env-purge-doc (app, env, docname)
+
+   :param app: :class:`.Sphinx`
+   :param env: :class:`.BuildEnvironment`
+   :param docname: ``str``
+
+   Emitted when all traces of a source file should be cleaned from the
+   environment, that is, if the source file is removed or before it is freshly read.
+   This is for extensions that keep their own caches
+   in attributes of the environment.
+
+   For example, there is a cache of all modules on the environment.
+   When a source file has been changed, the cache's entries for the file are cleared,
+   since the module declarations could have been removed from the file.
+
+   .. versionadded:: 0.5
+
+.. event:: env-before-read-docs (app, env, docnames)
+
+   :param app: :class:`.Sphinx`
+   :param env: :class:`.BuildEnvironment`
+   :param docnames: ``list[str]``
+
+   Emitted after the environment has determined the list of all added and
+   changed files and just before it reads them.
+   It allows extension authors to reorder
+   the list of docnames (*inplace*) before processing,
+   or add more docnames that Sphinx did not consider changed
+   (but never add any docnames that are not in ``env.found_docs``).
```

Comment:
Do we have an attribute reference for `found_docs`? If so, <code>:attr:\`env.found_docs <.Environment.found_docs>\`</code> may be a good idea.

Reply:
done